File: ts/indicators/joseph-granville/on-balance-volume/on-balance-volume.ts
import { buildMetadata } from '../../core/build-metadata';
import { Bar } from '../../../entities/bar';
import { BarComponent, barComponentValue } from '../../../entities/bar-component';
import { Scalar } from '../../../entities/scalar';
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { LineIndicator } from '../../core/line-indicator';
import { OnBalanceVolumeParams } from './params';

/** Function to calculate mnemonic of an __OnBalanceVolume__ indicator. */
export const onBalanceVolumeMnemonic = (params: OnBalanceVolumeParams = {}): string => {
  const suffix = componentTripleMnemonic(params.barComponent, params.quoteComponent, params.tradeComponent);
  if (suffix === '') {
    return 'obv';
  }
  return `obv(${suffix.slice(2)})`; // strip leading ", "
};

/**
 * OnBalanceVolume is Joseph Granville's On-Balance Volume (OBV).
 *
 * OBV is a cumulative volume indicator. On each update, if the price is higher
 * than the previous price, the volume is added to the running total; if the price
 * is lower, the volume is subtracted. If the price is unchanged, the total remains
 * the same.
 *
 * Reference:
 *
 * Granville, Joseph (1963). "Granville's New Key to Stock Market Profits".
 */
export class OnBalanceVolume extends LineIndicator {
  private previousSample: number;
  private value: number;
  private readonly barFunc: (bar: Bar) => number;

  /**
   * Constructs an instance given the parameters.
   */
  public constructor(params?: OnBalanceVolumeParams) {
    super();

    this.previousSample = 0;
    this.value = Number.NaN;
    this.primed = false;

    this.mnemonic = onBalanceVolumeMnemonic(params);
    this.description = 'On-Balance Volume OBV';

    // OBV defaults to ClosePrice, not TypicalPrice.
    const bc = params?.barComponent ?? BarComponent.Close;
    this.barFunc = barComponentValue(bc);
    this.barComponent = bc;
    this.quoteComponent = params?.quoteComponent;
    this.tradeComponent = params?.tradeComponent;
  }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return buildMetadata(
      IndicatorIdentifier.OnBalanceVolume,
      this.mnemonic,
      this.description,
      [
        { mnemonic: this.mnemonic, description: this.description },
      ],
    );
  }

  /** Updates the value of the indicator given the next sample (volume = 1). */
  public update(sample: number): number {
    return this.updateWithVolume(sample, 1);
  }

  /** Updates the value of the indicator given the next sample and volume. */
  public updateWithVolume(sample: number, volume: number): number {
    if (Number.isNaN(sample) || Number.isNaN(volume)) {
      return Number.NaN;
    }

    if (!this.primed) {
      this.value = volume;
      this.primed = true;
    } else {
      if (sample > this.previousSample) {
        this.value += volume;
      } else if (sample < this.previousSample) {
        this.value -= volume;
      }
    }

    this.previousSample = sample;

    return this.value;
  }

  /** Updates the indicator given the next bar sample, using bar volume. */
  public override updateBar(sample: Bar): IndicatorOutput {
    const price = this.barFunc(sample);
    const v = this.updateWithVolume(price, sample.volume);
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = v;
    return [scalar];
  }
}

File: go/indicators/josephgranville/onbalancevolume/onbalancevolume.go
package onbalancevolume

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
)

// OnBalanceVolume is Joseph Granville's On-Balance Volume (OBV).
//
// OBV is a cumulative volume indicator. On each update, if the price is higher
// than the previous price, the volume is added to the running total; if the price
// is lower, the volume is subtracted. If the price is unchanged, the total remains
// the same.
//
// Reference:
//
// Granville, Joseph (1963). "Granville's New Key to Stock Market Profits".
type OnBalanceVolume struct {
	mu sync.RWMutex
	core.LineIndicator
	barFunc        entities.BarFunc
	previousSample float64
	value          float64
	primed         bool
}

// NewOnBalanceVolume returns an instance of the indicator created using supplied parameters.
func NewOnBalanceVolume(p *OnBalanceVolumeParams) (*OnBalanceVolume, error) {
	const (
		invalid = "invalid on-balance volume parameters"
		fmtw    = "%s: %w"
	)

	bc := p.BarComponent
	if bc == 0 {
		bc = entities.BarClosePrice
	}

	qc := p.QuoteComponent
	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	tc := p.TradeComponent
	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	var (
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	mnemonic := "obv"
	if suffix := core.ComponentTripleMnemonic(bc, qc, tc); suffix != "" {
		mnemonic = "obv(" + suffix[2:] + ")" // strip leading ", "
	}

	desc := "On-Balance Volume OBV"

	o := &OnBalanceVolume{
		barFunc: barFunc,
		value:   math.NaN(),
	}

	// LineIndicator's Update uses volume=1 (scalar-only path).
	o.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, o.Update)

	return o, nil
}

// IsPrimed indicates whether the indicator is primed.
func (s *OnBalanceVolume) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes the output data of the indicator.
func (s *OnBalanceVolume) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.OnBalanceVolume,
		s.LineIndicator.Mnemonic,
		s.LineIndicator.Description,
		[]core.OutputText{
			{Mnemonic: s.LineIndicator.Mnemonic, Description: s.LineIndicator.Description},
		},
	)
}

// Update updates the indicator with the given sample using volume = 1.
// This satisfies the LineIndicator updateFn signature.
func (s *OnBalanceVolume) Update(sample float64) float64 {
	return s.UpdateWithVolume(sample, 1)
}

// UpdateWithVolume updates the indicator with the given sample and volume.
func (s *OnBalanceVolume) UpdateWithVolume(sample, volume float64) float64 {
	if math.IsNaN(sample) || math.IsNaN(volume) {
		return math.NaN()
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	if !s.primed {
		s.value = volume
		s.primed = true
	} else {
		if sample > s.previousSample {
			s.value += volume
		} else if sample < s.previousSample {
			s.value -= volume
		}
	}

	s.previousSample = sample

	return s.value
}

// UpdateBar updates the indicator given the next bar sample.
// This shadows LineIndicator.UpdateBar to use bar volume.
func (s *OnBalanceVolume) UpdateBar(sample *entities.Bar) core.Output {
	price := s.barFunc(sample)
	value := s.UpdateWithVolume(price, sample.Volume)

	output := make([]any, 1)
	output[0] = entities.Scalar{Time: sample.Time, Value: value}

	return output
}
